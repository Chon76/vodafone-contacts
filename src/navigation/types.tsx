/**
 * Learn more about using TypeScript with React Navigation:
 * https://reactnavigation.org/docs/typescript/
 */

import { NavigatorScreenParams } from '@react-navigation/native';
import { NativeStackScreenProps } from '@react-navigation/native-stack';

import { Contact } from '../types/Contact';

// to make useNavigation() or <Link /> typesafe
declare global {
  namespace ReactNavigation {
    interface RootParamList extends RootStackParamList {}
  }
}

export type RootStackParamList = {
  Home: NavigatorScreenParams<HomeStackParamList> | undefined;
  AddNewContact: undefined;
  NotFound: undefined;
};

export type HomeStackParamList = {
  ContactList: undefined;
  ContactDetails: { id: string };
  EditContact: { id: string; contact: Contact };
};

export type RootStackScreenProps<Screen extends keyof RootStackParamList> = NativeStackScreenProps<
  RootStackParamList,
  Screen
>;

export type HomeStackScreenProps<Screen extends keyof HomeStackParamList> = NativeStackScreenProps<
  HomeStackParamList,
  Screen
>;
