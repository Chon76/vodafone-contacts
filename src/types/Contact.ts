// "name": {
//     "title": "Miss",
//     "first": "Emmanuella",
//     "last": "Teunissen"
export type Contact = {
  name: {
    title: string;
    first: string;
    last: string;
  };
  picture: {
    large: string;
    medium: string;
    small: string;
  };
  phone: string;
  email: string;
  login: {
    uuid: string;
  };
};

// "results": [
//     {
//       "name": {
//         "title": "Mrs",
//         "first": "Glikeriya",
//         "last": "Tanskiy"
//       },
//       "email": "glikeriya.tanskiy@example.com",
//       "login": {
//         "uuid": "3d61bd2d-757e-4d12-bf69-068f305ded53",
//       "phone": "(066) J88-8313",
//       "picture": {
//         "large": "https://randomuser.me/api/portraits/women/19.jpg",
//         "medium": "https://randomuser.me/api/portraits/med/women/19.jpg",
//         "thumbnail": "https://randomuser.me/api/portraits/thumb/women/19.jpg"
//       },
//     }
//   ],
