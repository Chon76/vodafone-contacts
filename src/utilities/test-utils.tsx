/* eslint-disable import/no-extraneous-dependencies */
import { render, RenderOptions } from '@testing-library/react-native';
import React, { ReactElement } from 'react';
import { NativeBaseProvider } from 'native-base';
import { NavigationContainer } from '@react-navigation/native';

const inset = {
  frame: { x: 0, y: 0, width: 0, height: 0 },
  insets: { top: 0, left: 0, right: 0, bottom: 0 },
};

const wrapper = ({ children }: { children: React.ReactNode }) => {
  return (
    <NativeBaseProvider initialWindowMetrics={inset}>
      <NavigationContainer>{children}</NavigationContainer>
    </NativeBaseProvider>
  );
};

// wrap tests with necessery providers
const customRender = (ui: ReactElement, options?: Omit<RenderOptions, 'wrapper'>) =>
  render(ui, { wrapper, ...options });

export * from '@react-navigation/native';
export { customRender as render };
