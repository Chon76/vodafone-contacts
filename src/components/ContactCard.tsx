import { Box, HStack, Image, Text } from 'native-base';

import { Contact } from '../types/Contact';

type ContactCardProps = {
  contact: Contact;
};

export default function ContactCard({ contact }: ContactCardProps) {
  return (
    <HStack background={'background'} space={2} p={4} shadow={'4'} borderRadius="md">
      <Image
        source={{
          uri: contact.picture.large,
        }}
        alt={`${contact.name.first}-${contact.name.last}-avatar`}
        size="lg"
      />
      <Box flex={1}>
        <Text fontWeight={'bold'}>
          {contact.name.first},{contact.name.last}
        </Text>
        <Text>{contact.email}loremwpeepawerokrd</Text>
        <Text>{contact.phone}</Text>
      </Box>
    </HStack>
  );
}
