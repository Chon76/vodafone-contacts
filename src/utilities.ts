import { Contact } from './types/Contact';

// eslint-disable-next-line import/prefer-default-export
export function getInitialLeters(contacts: Contact[]): Set<string> {
  const initialLetters: Set<string> = new Set();

  contacts.forEach((contact) => {
    initialLetters.add(contact.name.first[0].toUpperCase());
    initialLetters.add(contact.name.last[0].toUpperCase());
  });

  console.log('initials', initialLetters);

  return initialLetters;
}
