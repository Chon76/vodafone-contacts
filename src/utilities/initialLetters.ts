import { Contact } from '../types/Contact';

/**
 *
 * @param contacts an array of {@link Contact}
 * @returns the initial letters of the names of the contact
 */
// eslint-disable-next-line import/prefer-default-export
export function getInitialLetters(contacts: Contact[]): Set<string> {
  const initialLetters: Set<string> = new Set();

  contacts.forEach((contact) => {
    initialLetters.add(contact.name.first[0].toUpperCase());
    initialLetters.add(contact.name.last[0].toUpperCase());
  });

  return initialLetters;
}
