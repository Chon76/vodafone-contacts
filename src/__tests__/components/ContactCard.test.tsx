import { v4 as uuidv4 } from 'uuid';

import ContactCard from '../../components/ContactCard';
import { Contact } from '../../types/Contact';

import { render } from '../../utilities/test-utils';

describe('contact card', () => {
  const contact: Contact = {
    id: uuidv4(),
    imageUrl: 'https://cdn.vodafone.co.uk/en/assets/images/large2X/IMG_Tech_Team_Tobi.jpg',

    name: {
      first: 'John',
      last: 'Doe',
    },
    email: 'john.doe@gmail.com',
    phone: '06 70 1234 567',
    address: '123 Sesame Street, Manhattan, New York',
  };

  it('renders correctly', () => {
    const header = render(<ContactCard contact={contact} />);

    expect(header.toJSON()).toMatchSnapshot();
  });
});
