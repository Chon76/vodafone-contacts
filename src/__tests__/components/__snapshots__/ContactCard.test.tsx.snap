// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`contact card renders correctly 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    [
      {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <View
    dataSet={{}}
    style={
      {
        "backgroundColor": "background",
        "borderRadius": 6,
        "elevation": 5,
        "marginLeft": 8,
        "marginRight": 8,
        "paddingBottom": 8,
        "paddingLeft": 8,
        "paddingRight": 8,
        "paddingTop": 8,
        "shadowColor": "#000000",
        "shadowOffset": {
          "height": 2,
          "width": 0,
        },
        "shadowOpacity": 0.25,
        "shadowRadius": 3.84,
      }
    }
  >
    <View
      dataSet={{}}
      style={
        {
          "flexDirection": "row",
        }
      }
    >
      <Image
        accessibilityLabel="John-Doe-avatar"
        alt="John-Doe-avatar"
        dataSet={{}}
        onError={[Function]}
        source={
          {
            "uri": "https://cdn.vodafone.co.uk/en/assets/images/large2X/IMG_Tech_Team_Tobi.jpg",
          }
        }
        style={
          {
            "height": 128,
            "maxWidth": "100%",
            "width": 128,
          }
        }
      />
      <View
        dataSet={{}}
        style={
          {
            "width": 8,
          }
        }
      />
      <View
        dataSet={{}}
        style={
          {
            "flex": 1,
          }
        }
      >
        <Text
          dataSet={{}}
          style={
            {
              "backgroundColor": undefined,
              "color": "#171717",
              "fontFamily": undefined,
              "fontSize": 16,
              "fontStyle": "normal",
              "fontWeight": "700",
              "letterSpacing": 0,
              "lineHeight": 24,
              "marginBottom": 4,
              "textAlign": "left",
              "textDecorationLine": undefined,
            }
          }
        >
          John
          ,
          Doe
        </Text>
        <View
          dataSet={{}}
          style={
            {
              "height": 8,
            }
          }
        />
        <Text
          dataSet={{}}
          style={
            {
              "backgroundColor": undefined,
              "color": "#3f3f46",
              "fontFamily": undefined,
              "fontSize": 14,
              "fontStyle": "normal",
              "fontWeight": "600",
              "letterSpacing": 0,
              "lineHeight": 21,
              "textDecorationLine": undefined,
            }
          }
        >
          john.doe@gmail.com
        </Text>
        <View
          dataSet={{}}
          style={
            {
              "height": 8,
            }
          }
        />
        <Text
          dataSet={{}}
          style={
            {
              "backgroundColor": undefined,
              "color": "#3f3f46",
              "fontFamily": undefined,
              "fontSize": 14,
              "fontStyle": "normal",
              "fontWeight": "600",
              "letterSpacing": 0,
              "lineHeight": 21,
              "textDecorationLine": undefined,
            }
          }
        >
          06 70 1234 567
        </Text>
      </View>
    </View>
  </View>
</RNCSafeAreaProvider>
`;
